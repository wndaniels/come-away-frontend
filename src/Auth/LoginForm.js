import Alert from "../Common/Alert.js";
import React, { useContext, useState } from "react";
import { useNavigate } from "react-router-dom";
import UserContext from "./UserContext.js";

/**
 * LOGIN FORM:
 *
 * Renders form and manages update to state on changes.
 *
 * On submission, calls login prop, navigates "/"
 *
 * Routes -> LoginForm -> Alert
 * Routed as /login
 *
 */

function LoginForm() {
  const navigate = useNavigate();
  const { login } = useContext(UserContext);
  const [formData, setFormData] = useState({
    username: "",
    password: "",
  });
  const [formErr, setFormErr] = useState([]);

  /**
   * Handle Form Submit:
   *
   * Calls login function prop and if successful, navigates to homepage.
   **/

  async function handleSubmit(evt) {
    evt.preventDefault();
    let res = await login(formData);
    if (res.success) {
      navigate("/");
    } else {
      setFormErr(res.errors);
    }
  }

  /**
   * Handles update for form data fields.
   */
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData((l) => ({ ...l, [name]: value }));
  };

  return (
    <div className="LoginForm m-5">
      <div className="container col-md-6 offset-md-3 col-lg-4 offset-lg-4">
        <h1 data-testid="header" className="mb-3">
          Login
        </h1>
        <div className="card">
          <div className="card-body">
            <form method="post" onSubmit={handleSubmit}>
              <div className="d-grid gap-3">
                <div className="form-group">
                  <input
                    type="text"
                    name="username"
                    placeholder="Username"
                    className="form-control"
                    onChange={handleChange}
                  />
                </div>
                <div className="form-group">
                  <input
                    type="password"
                    name="password"
                    placeholder="Password"
                    className="form-control mb-3"
                    onChange={handleChange}
                  />
                </div>
              </div>

              {formErr.length ? (
                <Alert type="danger" messages={formErr} />
              ) : null}

              <button onClick={handleSubmit} className="btn btn-sm btn-primary">
                Login
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  );
}

export default LoginForm;
