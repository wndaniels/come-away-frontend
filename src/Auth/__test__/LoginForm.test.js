import React, { useContext } from "react";
import LoginForm from "../LoginForm";
import { render, screen, getAllByText } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import "@testing-library/jest-dom";

jest.mock("../LoginForm");

it("renders without crashingt", async () => {
  const { asFragment } = await render(<LoginForm />);

  expect(asFragment(<LoginForm />)).toMatchSnapshot();
});

it("can find placeholder", async () => {
  await render(<LoginForm />);
  const usernamePlaceholder = screen.getAllByText(/"Username"/);
  console.log(usernamePlaceholder);
  expect(usernamePlaceholder).toBeInTheDocument();
});
// describe("LoginForm", () => {
//   it("renders without crashing", () => {
//     const mock = jest.fn(() => "I am a mock function");
//     expect(mock("Calling LoginForm")).toBe("I am a mock function");
//     expect(mock).toHaveBeenCalledWith("Calling LoginForm");
//   });

//   it("mock return value of a function one time", () => {
//     const mock = jest.fn();
//     mock.mockReturnValueOnce("Hello").mockReturnValueOnce("there!");
//     mock();
//     mock();

//     expect(mock).toHaveBeenCalledTimes(2);
//   });
// });

// it("should redirect user to homepage after login", () => {});

// describe("handle submit functionality", () => {
//   const handleSubmitMock = jest.fn();

//   test("should submit", () => {

//   })
// });
