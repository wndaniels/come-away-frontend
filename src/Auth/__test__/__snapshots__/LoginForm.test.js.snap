// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashingt 1`] = `<DocumentFragment />`;
